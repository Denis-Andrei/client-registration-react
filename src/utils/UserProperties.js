export let UserProperties =  {
    crn: "crn",
    name: "name",
    businessName: "businessName",
    contactNumber: "contactNumber",
    propertyNumber: "propertyNumber",
    postcode: "postcode",
    businessType: "businessType",
    password: "password",
    passwordCheck: "passwordCheck"
}

