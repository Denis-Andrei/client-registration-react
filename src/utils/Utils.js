


export const getData = (name) => sessionStorage.getItem(name) ? sessionStorage.getItem(name) : ""

export const isNameValid = (name) => /^([A-z])+(.?[a-zA-Z])*('?[a-zA-Z])*/.test(name)
export const isNumberValid = (number) => /^[0-9]{10}$|^[0-9]{11}$/.test(number) 
export const isPostcodeValid = (postcode) => /^(?:[A-Za-z]\d ?\d[A-Za-z]{2})|(?:[A-Za-z][A-Za-z\d]\d ?\d[A-Za-z]{2})|(?:[A-Za-z]{2}\d{2} ?\d[A-Za-z]{2})|(?:[A-Za-z]\d[A-Za-z] ?\d[A-Za-z]{2})|(?:[A-Za-z]{2}\d[A-Za-z] ?\d[A-Za-z]{2})/.test(postcode) 
export const isPasswordValid = (password) => password.length >= 10 
export const isPasswordMatching = (password, checkPassword) => password === checkPassword

