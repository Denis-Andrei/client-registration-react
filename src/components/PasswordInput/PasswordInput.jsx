import React, {useState} from 'react'
import { BackLink, H1, Label, LabelText, HintText, ErrorText, Input, FormGroup } from 'govuk-react'
import { Button } from 'govuk-react'
import { useHistory } from 'react-router-dom';
import { UserProperties } from '../../utils/UserProperties';
import { isPasswordMatching, isPasswordValid } from '../../utils/Utils';

export default function PasswordInput() {
    const history = useHistory();
     
    const [password, setPassword] = useState("");
    const [repeatPassword, setRepeatPassword] = useState("");
    const [hasPasswordErrors, setHasPasswordErrors] = useState(false)
    const [hasPasswordCheckErrors, setHasPasswordCheckErrors] = useState(false)


    const handleChangePass = (e) => {        
        setPassword(e.target.value)
    }
    
    const handleChangeRepeatPass = (e) => {        
        setRepeatPassword(e.target.value)
    }

    const handleSubmit = () =>{
        if(isPasswordValid(password) && isPasswordMatching(password, repeatPassword)){
            sessionStorage.setItem(UserProperties.password, password);
            sessionStorage.setItem(UserProperties.passwordCheck, repeatPassword);
            history.push('/summary')
        }else if(isPasswordMatching(password, repeatPassword)){
            setHasPasswordErrors(true);
        }else{
            setHasPasswordErrors(false);
            setHasPasswordCheckErrors(true)
        }
    } 

    return (
        <div>
            <BackLink data-testid="goBack" onClick={ () => history.goBack()}>Back</BackLink>
            <H1 data-testid="heading">Enter Password</H1>
            <FormGroup>
                <Label>
                    <LabelText data-testid="label-pass">Password</LabelText>
                    <HintText data-testid="hint-pass">Your password must be 10 or more characters. You can use a mix of letters, numbers or symbols.</HintText>
                    {
                        hasPasswordErrors 
                        ?
                        <ErrorText>Please provide a valid password</ErrorText>
                        :
                        ""
                    }
                    <Input data-testid="input-password" value={password} onChange={handleChangePass} type="password" />
                </Label>
                <Label>
                    <LabelText data-testid="label-re-enter-pass">Re-enter your password</LabelText>
                    <HintText data-testid="hint-re-enter-pass">Enter the same password as the one you entered above</HintText>
                    {
                        hasPasswordCheckErrors 
                        ?
                        <ErrorText>Your password should match</ErrorText>
                        :
                        ""
                    }
                    <Input data-testid="input-password-check" value={repeatPassword} onChange={handleChangeRepeatPass} type="password"/>
                </Label>
            </FormGroup>
            <Button data-testid="submit" onClick={handleSubmit}>Continue</Button>
        </div>
    )
}
