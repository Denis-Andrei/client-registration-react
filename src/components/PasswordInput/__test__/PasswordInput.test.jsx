import React from 'react';
import ReactDOM from 'react-dom';
import PasswordInput from '../PasswordInput';
import { render, cleanup, fireEvent } from '@testing-library/react'
import renderer from "react-test-renderer"
import * as Utils from '../../../utils/Utils'


afterEach(cleanup);

//Mocking history.push() and history.goBack()
const mockHistoryPush = jest.fn();
const mockGoBack = jest.fn();

    jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush,
        goBack: mockGoBack
    }),
}));
//end of mocking

it("PasswordInput() renders withoud crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<PasswordInput />, div)
})

it("renders PasswordInput() correctly", () => {
    const { getByTestId }  = render(<PasswordInput />)
    expect(getByTestId('heading')).toHaveTextContent("Enter Password")
    expect(getByTestId('label-pass')).toHaveTextContent("Password")
    expect(getByTestId('hint-pass')).toHaveTextContent("Your password must be 10 or more characters. You can use a mix of letters, numbers or symbols.")
    expect(getByTestId('label-re-enter-pass')).toHaveTextContent("Re-enter your password")
    expect(getByTestId('hint-re-enter-pass')).toHaveTextContent("Enter the same password as the one you entered above")
})

it("matches snapshot", () => {
    const tree = renderer.create(<PasswordInput />).toJSON();
    expect(tree).toMatchSnapshot();
})

describe("handleChangePass() ", () => {
    it("updates password input on change event", () => {
        const { getByTestId } = render(<PasswordInput />)
        const passwordInput = getByTestId("input-password");
    
        fireEvent.change(passwordInput, {target: {value: "test1test1"}})
        expect(passwordInput.value).toBe("test1test1")
    })
})

describe("handleChangeRepeatPass() ", () => {
    it("updates postcode input on change event", () => {
        const { getByTestId } = render(<PasswordInput />)
        const passwordCheckInput = getByTestId("input-password-check");
    
        fireEvent.change(passwordCheckInput, {target: {value: "test1test1"}})
        expect(passwordCheckInput.value).toBe("test1test1")
    })
})

describe("handleSubmit() ", () => {
    it("onSubmit redirects to the correct route", () => {
        
        const mockIsPasswordValid =  jest.fn().mockReturnValue(true)
        Utils.isPasswordValid = mockIsPasswordValid
        
        const mockIsPasswordMatching =  jest.fn().mockReturnValue(true)
        Utils.isPasswordMatching = mockIsPasswordMatching

        const { getByTestId } = render(<PasswordInput />)
        const submit = getByTestId("submit");

        fireEvent.click(submit);
        expect(mockHistoryPush).toHaveBeenCalledWith('/summary');
    })
})

describe("goBack() ", () => {
    it("should be called on click event", () => {
        const { getByTestId } = render(<PasswordInput />)
        const goBackBtn = getByTestId("goBack");
    
        fireEvent.click(goBackBtn)
        expect(mockGoBack).toHaveBeenCalled();
    })
})
