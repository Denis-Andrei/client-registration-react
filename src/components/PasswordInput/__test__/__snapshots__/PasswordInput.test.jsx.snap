// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <a
    className="src__Anchor-sc-11edswm-0 gGueb"
    data-testid="goBack"
    onClick={[Function]}
  >
    Back
  </a>
  <h1
    className="heading__StyledHeading-sc-1sovo83-0 juamsS"
    data-testid="heading"
    size="XLARGE"
  >
    Enter Password
  </h1>
  <div
    className="src__FormGroup-kqurat-0 fREKLF"
  >
    <label
      className="src__StyledLabel-sc-14n2scx-0 kHdLXy"
    >
      <span
        className="src__StyledLabelText-sc-12l0n4z-0 HDdRl"
        data-testid="label-pass"
      >
        Password
      </span>
      <span
        className="src__StyledHint-sc-1qn4br2-0 kngyqG"
        data-testid="hint-pass"
      >
        Your password must be 10 or more characters. You can use a mix of letters, numbers or symbols.
      </span>
      
      <input
        className="src__StyledInput-u8exb4-0 iqwpbG"
        data-testid="input-password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
    <label
      className="src__StyledLabel-sc-14n2scx-0 kHdLXy"
    >
      <span
        className="src__StyledLabelText-sc-12l0n4z-0 HDdRl"
        data-testid="label-re-enter-pass"
      >
        Re-enter your password
      </span>
      <span
        className="src__StyledHint-sc-1qn4br2-0 kngyqG"
        data-testid="hint-re-enter-pass"
      >
        Enter the same password as the one you entered above
      </span>
      
      <input
        className="src__StyledInput-u8exb4-0 iqwpbG"
        data-testid="input-password-check"
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
  </div>
  <button
    className="sc-bdnxRM gXVtdA"
    data-testid="submit"
    disabled={false}
    onClick={[Function]}
  >
    Continue
  </button>
</div>
`;
